### PR TITLE
[fp8]Moe support fp8 communication

### LINKED ISSUE #5954: support moe

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

closed #5954

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Moe support fp8 communication

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by flybird11111** [comment]
image
test deepseek finetune.

**Comment by flybird11111** [comment]
image
test mixtral example.

File: colossalai/quantization/fp8.py [hunk 1]
```diff
@@ -27,16 +27,21 @@ def cast_to_fp8(inp: torch.Tensor, fp8_format="e4m3", per_channel_scale=False) -
...
-        per_channel_max = torch.where(per_channel_max > 0, per_channel_max, 1.0)
-        scale = fp8_max / per_channel_max[:, None]
-        scale_inv = per_channel_max / fp8_max
-    else:
-        per_tensor_max = inp.abs().max().float()
-        per_tensor_max = torch.where(per_tensor_max > 0, per_tensor_max, 1.0)
+    if inp.numel() == 0:
+        per_tensor_max = torch.tensor([1.0], device=inp.device)
         scale = fp8_max / per_tensor_max
         scale_inv = 1.0 / scale
```
Comment (ver217) [inline colossalai/quantization/fp8.py:33]: No need to compute amax if the tensor is empty

File: colossalai/shardformer/shard/sharder.py [hunk 2]
```diff
@@ -198,11 +198,12 @@ def _replace_sub_module(
...
-                    f"Failed to replace {suffix} of type {native_sub_module.__class__.__qualname__}"
-                    f" with {target_module.__qualname__} with the exception: {e}. "
-                    "Please check your model configuration or sharding policy, you can set up an issue for us to help you as well."
-                )
+                # raise RuntimeError(
+                #     f"Failed to replace {suffix} of type {native_sub_module.__class__.__qualname__}"
+                #     f" with {target_module.__qualname__} with the exception: {e}. "
+                #     "Please check your model configuration or sharding policy, you can set up an issue for us to help you as well."
+                # )
+                raise e
```
Comment (ver217) [inline colossalai/shardformer/shard/sharder.py:206]: This affects try-catch block of previous level code. Better don't modify

**Review by ver217 — APPROVED** [review]

**Comment by flybird11111** [comment]
image
update mixtral test screenshots.

**Review by GuangyaoZhang — APPROVED** [review]

**Comment by flybird11111** [comment]
update deepseek test screenshot

**Pull request merged**